### ISSUE TITLE
ParamSpec parameters don't match __init__ parameters of class 

### ISSUE THREAD

**Issue by NeilGirdhar** [python/mypy#12986]
```python
from functools import partial
from typing import Callable, Generic, ParamSpec

P = ParamSpec('P')

class custom_vjp(Generic[P]):
    def __init__(self,
                 fun: Callable[P, None],
                 x: int = 0):
        super().__init__()
        self.fun = fun

    def __call__(self, *args: P.args, **kwargs: P.kwargs) -> None:
        return self.fun(*args, **kwargs)

@partial(custom_vjp, x=1)  # error: Argument 1 to "partial" has incompatible type "Type[custom_vjp[Any]]"; expected "Callable[..., custom_vjp[P]]"
def f() -> None:
    pass
```

**Comment by kstauffer** [comment]
You get the same error without the decorator (delete the `@` character and the last two lines entirely). The error is at the instantiation of the `partial` object.

The problem is the use of an unbound type in a way that mypy cannot handle. Where the error message says "Any" is misleading; the internal state of mypy still has "P" unbound. Notice that mypy is happy if you explicitly annotate or cast with the same dubious type mypy already claims to think it has:

```
con: Type[custom_vjp[Any]] = custom_vjp
partial(con, x=1)  # no issues found
partial(cast(Type[custom_vjp[Any]], custom_vjp), x=1)  # no issues found
```

That's proof that the mypy error message is not really telling you what's gone wrong inside mypy.

I am working on a plugin to properly handle functools.partial and partialmethod. I'll be sure this handle this test case.

**Comment by NeilGirdhar** [comment]
Awesome!! Really looking forward to your plugin.  And thank you for diagnosing the error and providing a workaround.

**Comment by ilevkivskyi** [comment]
The original example works on master, likely fixed by https://github.com/python/mypy/pull/15837
